File: C#/TransportNetworking/transport/network/rpcprotocol/TransportClientRpcReflectionWorker.cs
using System.Net.Sockets;
using System.Reflection;
using System.Runtime.Serialization.Formatters.Binary;
using TransportServices.transport.services;
using JsonSerializer = System.Text.Json.JsonSerializer;
using Newtonsoft.Json;
using Transport.Services;
using TransportModel.transport.model;

namespace TransportNetworking.transport.network.rpcprotocol;

public class TransportClientRpcReflectionWorker : ITransportObserver
{
    private ITransportServices server;
    private Socket connection;

    private StreamReader input;
    private StreamWriter output;
    private volatile bool connected;

    public TransportClientRpcReflectionWorker(ITransportServices server, Socket connection)
    {
        this.server = server;
        this.connection = connection;
        try
        {
            NetworkStream networkStream = new NetworkStream(connection);
            output = new StreamWriter(networkStream);
            input = new StreamReader(networkStream);
            connected = true;
        }
        catch (IOException e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }

    public void Run()
    {
        while (connected)
        {
            try
            {
                string line = input.ReadLine();
                Console.WriteLine(line);
                Request request = JsonConvert.DeserializeObject<Request>(line!);
                //BinaryFormatter formatter = new BinaryFormatter();
                //object request = formatter.Deserialize(input.BaseStream);
                Console.WriteLine("RUN:" + request);
                Response response = HandleRequest(request);
                if (response != null)
                {
                    SendResponse(response);
                }
            }
            catch (IOException e)
            {
                Console.WriteLine(e.StackTrace);
            }

            Thread.Sleep(1000);
        }

        try
        {
            input.Close();
            output.Close();
            connection.Close();
        }
        catch (IOException e)
        {
            Console.WriteLine("Error " + e);
        }
    }
    
    
    private void SendResponse(Response response)
    {
        //Console.WriteLine("sending response " + response);
        lock (output)
        {
            string responseString = JsonConvert.SerializeObject(response);
            output.WriteLine(responseString);
            output.Flush();
            //BinaryFormatter formatter = new BinaryFormatter();
            /*using (MemoryStream ms = new MemoryStream())
            {
                formatter.Serialize(ms, response);
                byte[] buffer = ms.ToArray();
                output.Write(buffer, 0, buffer.Length);
                output.Flush();
            }*/
        }
    }

    
    //asta ii pentru update
    public void ReservationAdded(Object obj)
    {
        Response resp = new Response.Builder().Type(ResponseType.TICKET_ADDED_NOTIFICATION).Data(obj).Build();
        Console.WriteLine("Reservation added: " + obj);
        try
        {
            SendResponse(resp);
        }
        catch (IOException e)
        {
            Console.WriteLine(e.StackTrace);
        }
    }

    private static Response okResponse = new Response.Builder().Type(ResponseType.OK).Build();

    private Response HandleRequest(Request request)
    {
        Response response = null;
        string handlerName = "Handle" + request.Type;
        Console.WriteLine("HandlerName " + handlerName);
        try
        {
            MethodInfo method = this.GetType().GetMethod(handlerName, new Type[] { typeof(Request) });
            if (method != null)
            {
                response = (Response)method.Invoke(this, new object[] { request });
                //Console.WriteLine("Method " + handlerName + " invoked");
            }
            else
            {
                Console.WriteLine("Method " + handlerName + " not found");
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
        }

        return response;
    }

    public Response HandleGET_ALL_FLIGHTS(Request request)
    {
        Console.WriteLine("Handling GET_ALL_FLIGHTS");
        
        try
        {
            //Console.WriteLine("Trimit avioanele");
            return new Response.Builder().Type(ResponseType.GET_ALL_FLIGHTS).Data(server.GetAllFlights()).Build();
        }
        catch (TransportException e)
        {
            return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
        }
        
        
    }

    public Response HandleLOGIN(Request request)
    {
        Console.WriteLine("Login Request ... " + request.Type);
        
        if (request.Data == null)
        {
            return new Response.Builder().Type(ResponseType.ERROR).Data("Request data is null").Build();
        }

        User user = request.Data as User; 
        if (user == null)
        {

            return new Response.Builder().Type(ResponseType.ERROR).Data("Invalid user data").Build();
        }

        try
        {
            //Console.WriteLine("in HandleLOGIN: ");
            Boolean succes = server.Login(user.Username, user.Password, this);

            if (succes == true){
                Console.WriteLine("Login successful! (Worker)");
                return okResponse;
            }
            else
            {
                return new Response.Builder().Type(ResponseType.ERROR).Data("").Build();
            }
        }
        catch (Exception e)
        {
            Console.WriteLine(e.StackTrace);
            connected = false;
            return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
        }

    }

    public Response HandleSEARCH_FLIGHT_BY_DESTINATION_DATE(Request request)
    {
        Object[] data = (Object[])request.Data;
        string destination = (string) data[0];
        DateTime departureDate = (DateTime) data[1];
        
        try {
            List<Flight> flightDetails = server.SearchFlights(destination, departureDate);
            return new Response.Builder().Type(ResponseType.SEARCH_FLIGHT_BY_DESTINATION_DATE).Data(flightDetails).Build();
        } catch (TransportException e) {
            return new Response.Builder().Type(ResponseType.ERROR).Data(e.Message).Build();
        }
    }

    public Response HandleNEW_TICKET(Request request)
    {
        String data = (string)request.Data;

        List<string> obj = data.Split(';')
            .Select(s => s.Trim())
            .ToList();
        
        List<string> passagersList = obj[3].Split(',')
            .Select(s => s.Trim())
            .ToList();

        Boolean success = server.BuyTicket(obj[0], int.Parse(obj[1]),int.Parse(obj[2]), passagersList);

        if (success) {
            return okResponse;
        } else {
            return new Response.Builder()
                .Type(ResponseType.ERROR)
                .Data("Something went bad.")
                .Build();
        }
    }
}

File: Java/TransportNetworking/src/main/java/transport/network/rpcprotocol/TransportClientRpcReflectionWorker.java
package transport.network.rpcprotocol;

import transport.model.Flight;
import transport.model.Ticket;
import transport.model.User;
import transport.services.*;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;

public class TransportClientRpcReflectionWorker implements Runnable, ITransportObserver {
    private ITransportServices server;
    private Socket connection;
    ITransportObserver client;

    private ObjectInputStream input;
    private ObjectOutputStream output;
    private volatile boolean connected;

    public TransportClientRpcReflectionWorker(ITransportServices server, Socket connection) {
        System.out.println("IN TRANSPORT WORKER CONSTRUCTOR");
        this.server = server;
        this.connection = connection;
        try {
            if (connection != null) {
                System.out.println("IN TRANSPORT WORKER CONNECTION conectiune nenula");
                output = new ObjectOutputStream(connection.getOutputStream());
                output.flush();
                input = new ObjectInputStream(connection.getInputStream());
                connected = true;
            } else {
                System.out.println("IN TRANSPORT WORKER CONNECTION conectiune nenula");
                throw new IOException("Socket connection is null.");
            }
        } catch (IOException e) {
            e.printStackTrace();
            connected = false;
        }
    }

    public void run() {
        System.out.println("IN TRANSPORT WORKER RUN");
        while (connected) {
            System.out.println("IN TRANSPORT WORKER RUN");
            try {
                Object request = input.readObject();
                System.out.println("Response primit din handleRequest.... ");
                Response response = handleRequest((Request) request);
                System.out.println("Response primit din handleRequest: " + response);
                if (response != null) {
                    sendResponse(response);
                }
            } catch (IOException e) {
                e.printStackTrace();
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error " + e);
        }
    }

    private static Response okResponse = new Response.Builder().type(ResponseType.OK).build();

    private Response handleRequest(Request request) {
        System.out.println("IN TRANSPORT WORKER HANDLE REQUEST");
        Response response = null;
        String handlerName = "handle" + (request).type();
        System.out.println("HandlerName " + handlerName);
        try {
            Method method = this.getClass().getDeclaredMethod(handlerName, Request.class);
            response = (Response) method.invoke(this, request);
            System.out.println("Method " + handlerName + " invoked");
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (InvocationTargetException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        }

        return response;
    }

    private void sendResponse(Response response) throws IOException {
        System.out.println("IN TRANSPORT WORKER SEND RESPONSE");
        System.out.println("sending response " + response);
        synchronized (output) {
            output.writeObject(response);
            output.flush();
        }
    }

    private Response handleLOGIN(Request request) {
        System.out.println("Login Request ... " + request.type());
        User user = (User) request.data();
        try {
            server.login(user.getUsername(), user.getPassword(), this);
            return okResponse;
        } catch (TransportException e) {
            connected=false;
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }


    private Response handleGET_ALL_FLIGHTS(Request request) {
        System.out.println("Get All Trips Request ...");
        try {
            List<Flight> allFlights = server.getAllFlights();
            return new Response.Builder().type(ResponseType.GET_ALL_FLIGHTS).data(allFlights).build();
        } catch (TransportException e) {
            e.printStackTrace();
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }

    }

    private Response handleSEARCH_FLIGHT_BY_DESTINATION_DATE(Request request) {
        System.out.println("Search flights by destination and date request...");

        Object[] data = (Object[]) request.data();
        String destination = (String) data[0];
        LocalDate departureDate = (LocalDate) data[1];

        try {
            List<String> flightDetails = server.searchFlights(destination, departureDate);  // Ar trebui să returneze deja lista de String-uri
            return new Response.Builder().type(ResponseType.SEARCH_FLIGHT_BY_DESTINATION_DATE).data(flightDetails).build();
        } catch (TransportException e) {
            return new Response.Builder().type(ResponseType.ERROR).data(e.getMessage()).build();
        }
    }


    private Response handleLOGOUT(Request request) {
        System.out.println("Logout Request...");
        User user = (User) request.data();
        try {
            server.logout(user.getUsername(), this);
            connected = false;
            return okResponse;
        } catch (TransportException e) {
            return new Response.Builder()
                    .type(ResponseType.ERROR)
                    .data(e.getMessage())
                    .build();
        }
    }

    private Response handleNEW_TICKET(Request request) {
        System.out.println("Buy Ticket Request... worker");
        String data = (String) request.data();

        List<String> obj = Arrays.stream(data.split(";"))
                .map(String::trim)
                .collect(Collectors.toList());

        System.out.println(data);

        List<String> passagersList = Arrays.stream(obj.get(2).split(","))
                .map(String::trim)
                .collect(Collectors.toList());

        System.out.println(obj.get(0));
        System.out.println(passagersList);
        System.out.println(obj.get(1));

        boolean success = server.buyTicket(obj.get(0), Integer.parseInt(obj.get(1)), passagersList);

        if (success) {
            return okResponse;
        } else {
            return new Response.Builder()
                    .type(ResponseType.ERROR)
                    .data("Something went bad.")
                    .build();
        }


    }

    public void handleTICKET_ADDED_NOTIFICATION(Request request) {
        Response resp = new Response.Builder()
                .type(ResponseType.TICKET_ADDED_NOTIFICATION)
                .data(this)
                .build();
        try {
            sendResponse(resp);
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void handleCLIENT(Request request) {
        ITransportObserver observer = (ITransportObserver) request.data();
        this.client = observer;

        Response resp = new Response.Builder()
                .type(ResponseType.CLIENT)
                .data("ok")
                .build();
        try {
            sendResponse(resp);
        }
        catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public void reservationAdded(Object object) {
        Response resp = new Response.Builder().type(ResponseType.TICKET_ADDED_NOTIFICATION).data(object).build();
        System.out.println("Reservation added: ");
        try {
            sendResponse(resp);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

}
